### PR TITLE
Add MDD Dataset

### PR DESCRIPTION
**Pull request by gchhablani**
- **Name:** *MDD Dataset*
- **Description:** The Movie Dialog dataset (MDD) is designed to measure how well models can perform at goal and non-goal orientated dialog centered around the topic of movies (question answering, recommendation and discussion), from various movie reviews sources such as MovieLens and OMDb.
- **Paper:** [arXiv](https://arxiv.org/pdf/1511.06931.pdf)
- **Data:**  https://research.fb.com/downloads/babi/
- **Motivation:** This is one of the popular dialog datasets, a part of Facebook Research's "bAbI project".

### Checkbox

- [x] Create the dataset script `/datasets/my_dataset/my_dataset.py` using the template
- [x] Fill the `_DESCRIPTION` and `_CITATION` variables
- [x] Implement `_infos()`, `_split_generators()` and `_generate_examples()`
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is filled with the different configurations of the dataset and that the `BUILDER_CONFIG_CLASS` is specified if there is a custom config class.
- [x] Generate the metadata file `dataset_infos.json` for all configurations
- [x] Generate the dummy data `dummy_data.zip` files to have the dataset script tested and that they don't weigh too much (<50KB)
- [x] Add the dataset card `README.md` using the template : fill the tags and the various paragraphs
- [x] Both tests for the real data and the dummy data pass.

**Note**: I haven't included the following from the data files: `entities` (the file containing list of all entities in the first three subtasks), `dictionary`(the dictionary of words they use in their models), `movie_kb`(contains the knowledge base of information about the movies, actors and other entities that are mentioned in the dialogs). Please let me know if those are needed, and if yes, should I make separate configurations for them?

### CODE REVIEW

File: datasets/mdd/mdd.py [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+    DEFAULT_CONFIG_NAME = (
+        "task1_qa"  # It's not mandatory to have a default configuration. Just use one if it make sense.
+    )
+
+    def _info(self):
+        features = datasets.Features(
+            {
+                "dialogue_turns": datasets.Sequence(
+                    {
+                        "speaker": datasets.ClassLabel(names=["0", "1"]),
```
Comment (lhoestq) [inline datasets/mdd/mdd.py:444]: You can just have an integer here I think, "speaker" is not a label we want to predict

File: datasets/mdd/mdd.py [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+                        "filepath": os.path.join(data_dir, paths[self.config.name]["cand_test"]),
+                    },
+                ),
+            ]
+        return splits
+
+    def _generate_examples(self, filepath):
+        if "cand" not in filepath:
+            with open(filepath, encoding="utf-8") as f:
+                data = f.readlines()
```
Comment (lhoestq) [inline datasets/mdd/mdd.py:519]: can you iterate over the lines instead of reading the full file to save memory ?

File: datasets/mdd/mdd.py [hunk 3]
```diff
@@ -0,0 +1,569 @@
...
+                            yield example_idx, {"dialogue_turns": dialogue_turns}
+                    else:
+                        exchange = line[len(line.split()[0]) :].strip().split("\t")  # Skip the number in the front
+                        sp1 = exchange[0]
+                        sp2 = exchange[-1]  # Might contain multiple tabs in between.
+                        dialogue_turns.append({"speaker": 0, "utterance": sp1})
+                        dialogue_turns.append({"speaker": 1, "utterance": sp2})
+        else:
+            with open(filepath, encoding="utf-8") as f:
+                data = f.readlines()
```
Comment (lhoestq) [inline datasets/mdd/mdd.py:547]: same here

File: datasets/mdd/README.md [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+## Dataset Structure
+
+### Data Instances
+An instance from the `task3_qarecs` config's `train` split:
+```
+{'dialogue_turns': {'speaker': [0, 1, 0, 1, 0, 1], 'utterance': ["I really like Jaws, Bottle Rocket, Saving Private Ryan, Tommy Boy, The Muppet Movie, Face/Off, and Cool Hand Luke. I'm looking for a Documentary movie.", 'Beyond the Mat', 'Who is that directed by?', 'Barry W. Blaustein', 'I like Jon Fauer movies more. Do you know anything else?', 'Cinematographer Style']}}
+```
+An instance from the `task4_reddit` config's `cand-valid` split:
+```
+{'dialogue_turns': {'speaker': [0], 'utterance': ['MORTAL KOMBAT !']}}
```
Comment (lhoestq) [inline datasets/mdd/README.md:91]: This is indeed a dialogue worthy of reddit xD

**Review by lhoestq — COMMENTED** [review]
Nice thanks :)

File: datasets/mdd/mdd.py [hunk 5]
```diff
@@ -0,0 +1,569 @@
...
+        datasets.BuilderConfig(
+            name="task4_reddit",
+            version=VERSION,
+            description="This part of my dataset covers task4_reddit part of the dataset",
+        ),
+    ]
+
+    DEFAULT_CONFIG_NAME = (
+        "task1_qa"  # It's not mandatory to have a default configuration. Just use one if it make sense.
+    )
```
[the remark this answers is not in the document]
Comment (gchhablani) [reply datasets/mdd/mdd.py:437]: Removed.

File: datasets/mdd/README.md (hunk 4 above)
Comment (gchhablani) [reply datasets/mdd/README.md:91]: xD Hahaha

**Comment by gchhablani** [comment]
Hi @lhoestq,

I have added changes from review.

**Review by lhoestq — APPROVED** [review]
LGTM !

**Comment by gchhablani** [comment]
Thanks for approving :)

**Pull request merged**